lib/kafka_ex/server.ex

```diff
@@ -80,10 +95,23 @@ defmodule KafkaEx.Server do
     Process.exit(event_pid, :kill)
   end
 
-  defp start_stream(pid, metadata, client, handler, topic, partition, request_fn) do
-    {metadata, client, {:ok, response}} = send_request(metadata, client, topic, partition, request_fn, &KafkaEx.Protocol.Fetch.parse_response/1)
-    response[topic][partition][:message_set] |>
-      Enum.each(fn(message_set) -> GenEvent.notify(pid, message_set) end)
-    start_stream(pid, metadata, client, handler, topic, partition, request_fn)
+  defp start_stream(pid, metadata, client, handler, topic, partition, offset) do
+    request_fn = KafkaEx.Protocol.Fetch.create_request_fn(topic, partition, offset, @wait_time, @min_bytes, @max_bytes)
+    {metadata, client} = send_request(metadata, client, topic, partition, request_fn, nil, 100, false)
+    receive do
+      {:tcp, _, data} ->
+        {:ok, response} = KafkaEx.Protocol.Fetch.parse_response(data)
+        message_sets = response[topic][partition].message_set
+        if Enum.empty?(message_sets) do
+          :timer.sleep(1000)
+          start_stream(pid, metadata, client, handler, topic, partition, offset)
+        else
+          Enum.each(message_sets, fn(message_set) -> GenEvent.notify(pid, message_set) end)
+          last_offset = Enum.at(Enum.reverse(message_sets), 0).offset
```

Comment:
I think we can abstract the logic here into a recursive function smartly returns the last_offset

e.g:

```elixir
defp notify_handler([], handler_pid, offset), do: offset

defp notify_handler([message_set|message_sets], handler_pid, offset) do
   GenEvent.notify(handler_pid, message_set)
   notify_handler(message_sets, handler_pid, message_set.offset)
end
```

that saves us the cost of doing `Enum.reverse` and we can conveniently remove the `if/else` block